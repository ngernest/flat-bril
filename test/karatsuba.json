{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "result",
          "funcs": [
            "karatsuba"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "base",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "x"
          ],
          "dest": "x_digits",
          "funcs": [
            "num_digits"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "y_digits",
          "funcs": [
            "num_digits"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x_digits",
            "one"
          ],
          "dest": "x_1_digit",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "y_digits",
            "one"
          ],
          "dest": "y_1_digit",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "x_1_digit",
            "y_1_digit"
          ],
          "dest": "cond",
          "op": "or",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "result",
          "op": "mul",
          "type": "int"
        },
        {
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "x_digits",
            "y_digits"
          ],
          "dest": "n",
          "funcs": [
            "max"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "n2",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "base",
            "n2"
          ],
          "dest": "pow_n2",
          "funcs": [
            "pow"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x",
            "pow_n2"
          ],
          "dest": "a",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "x",
            "pow_n2"
          ],
          "dest": "b",
          "funcs": [
            "mod"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "y",
            "pow_n2"
          ],
          "dest": "c",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "y",
            "pow_n2"
          ],
          "dest": "d",
          "funcs": [
            "mod"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "b",
            "d"
          ],
          "dest": "z0",
          "funcs": [
            "karatsuba"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "a_plus_b",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "c",
            "d"
          ],
          "dest": "c_plus_d",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "a_plus_b",
            "c_plus_d"
          ],
          "dest": "z1",
          "funcs": [
            "karatsuba"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a",
            "c"
          ],
          "dest": "z2",
          "funcs": [
            "karatsuba"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n2",
            "two"
          ],
          "dest": "hi_shift_factor",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "base",
            "hi_shift_factor"
          ],
          "dest": "hi_shift",
          "funcs": [
            "pow"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "z2",
            "hi_shift"
          ],
          "dest": "hi_shifted",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "z1",
            "z2"
          ],
          "dest": "mid",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "mid",
            "z0"
          ],
          "dest": "mid",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "mid",
            "pow_n2"
          ],
          "dest": "mid_shifted",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "hi_shifted",
            "mid_shifted"
          ],
          "dest": "result",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "result",
            "z0"
          ],
          "dest": "result",
          "op": "add",
          "type": "int"
        },
        {
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "result"
          ],
          "op": "ret"
        }
      ],
      "name": "karatsuba",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "base",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "one"
          ],
          "dest": "digits",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "one"
          ],
          "dest": "power",
          "op": "id",
          "type": "int"
        },
        {
          "label": "while.cond"
        },
        {
          "args": [
            "n",
            "power"
          ],
          "dest": "temp",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "temp",
            "base"
          ],
          "dest": "is_term",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "is_term"
          ],
          "labels": [
            "while.finish",
            "while.body"
          ],
          "op": "br"
        },
        {
          "label": "while.body"
        },
        {
          "args": [
            "digits",
            "one"
          ],
          "dest": "digits",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "power",
            "base"
          ],
          "dest": "power",
          "op": "mul",
          "type": "int"
        },
        {
          "labels": [
            "while.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "while.finish"
        },
        {
          "args": [
            "digits"
          ],
          "op": "ret"
        }
      ],
      "name": "num_digits",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "cond",
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "a"
          ],
          "op": "ret"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "b"
          ],
          "op": "ret"
        }
      ],
      "name": "max",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "c",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "c",
            "b"
          ],
          "dest": "d",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a",
            "d"
          ],
          "dest": "remainder",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "remainder"
          ],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "v0",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0"
          ],
          "dest": "res",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v2"
          ],
          "dest": "_i",
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "_i"
          ],
          "dest": "v3",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v4",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v5",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v4",
            "v5"
          ],
          "dest": "v6",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v6"
          ],
          "dest": "v7",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v7"
          ],
          "labels": [
            "for.body.1",
            "for.end.1"
          ],
          "op": "br"
        },
        {
          "label": "for.body.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v8",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v9",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v10"
          ],
          "dest": "res",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "_i"
          ],
          "dest": "v11",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v12",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v11",
            "v12"
          ],
          "dest": "v13",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v13"
          ],
          "dest": "_i",
          "op": "id",
          "type": "int"
        },
        {
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v14",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v14"
          ],
          "op": "ret"
        }
      ],
      "name": "pow",
      "type": "int"
    }
  ]
}
