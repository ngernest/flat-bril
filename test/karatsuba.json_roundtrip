{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "result",
          "funcs": [
            "karatsuba"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "base",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "x"
          ],
          "dest": "x_digits",
          "funcs": [
            "num_digits"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "y_digits",
          "funcs": [
            "num_digits"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x_digits",
            "one"
          ],
          "dest": "x_1_digit",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "y_digits",
            "one"
          ],
          "dest": "y_1_digit",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "x_1_digit",
            "y_1_digit"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "or",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "funcs": [],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "x_digits",
            "y_digits"
          ],
          "dest": "n",
          "funcs": [
            "max"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "n2",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "base",
            "n2"
          ],
          "dest": "pow_n2",
          "funcs": [
            "pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x",
            "pow_n2"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "x",
            "pow_n2"
          ],
          "dest": "b",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "y",
            "pow_n2"
          ],
          "dest": "c",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "y",
            "pow_n2"
          ],
          "dest": "d",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "b",
            "d"
          ],
          "dest": "z0",
          "funcs": [
            "karatsuba"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "a_plus_b",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "c",
            "d"
          ],
          "dest": "c_plus_d",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "a_plus_b",
            "c_plus_d"
          ],
          "dest": "z1",
          "funcs": [
            "karatsuba"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a",
            "c"
          ],
          "dest": "z2",
          "funcs": [
            "karatsuba"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n2",
            "two"
          ],
          "dest": "hi_shift_factor",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "base",
            "hi_shift_factor"
          ],
          "dest": "hi_shift",
          "funcs": [
            "pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "z2",
            "hi_shift"
          ],
          "dest": "hi_shifted",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "z1",
            "z2"
          ],
          "dest": "mid",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "mid",
            "z0"
          ],
          "dest": "mid",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "mid",
            "pow_n2"
          ],
          "dest": "mid_shifted",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "hi_shifted",
            "mid_shifted"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "result",
            "z0"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "karatsuba",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "base",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "one"
          ],
          "dest": "digits",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "one"
          ],
          "dest": "power",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "while.cond"
        },
        {
          "args": [
            "n",
            "power"
          ],
          "dest": "temp",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "temp",
            "base"
          ],
          "dest": "is_term",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "is_term"
          ],
          "funcs": [],
          "labels": [
            "while.finish",
            "while.body"
          ],
          "op": "br"
        },
        {
          "label": "while.body"
        },
        {
          "args": [
            "digits",
            "one"
          ],
          "dest": "digits",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "power",
            "base"
          ],
          "dest": "power",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "while.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "while.finish"
        },
        {
          "args": [
            "digits"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "num_digits",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "funcs": [],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "a"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "b"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "max",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "c",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "c",
            "b"
          ],
          "dest": "d",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a",
            "d"
          ],
          "dest": "remainder",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "remainder"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v2"
          ],
          "dest": "_i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "_i"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v5",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v4",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v6"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v7"
          ],
          "funcs": [],
          "labels": [
            "for.body.1",
            "for.end.1"
          ],
          "op": "br"
        },
        {
          "label": "for.body.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v10"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "_i"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v12",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v11",
            "v12"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v13"
          ],
          "dest": "_i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v14"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "pow",
      "type": "int"
    }
  ]
}
