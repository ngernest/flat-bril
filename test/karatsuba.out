@main
	op: call 		dest: result		type: int		args: ["x", "y"]		funcs: ["karatsuba"]
	op: print		args: ["result"]	
@karatsuba
	op: const		dest: one		type: int		value: 1	
	op: const		dest: two		type: int		value: 2	
	op: const		dest: base		type: int		value: 10	
	op: call 		dest: x_digits		type: int		args: ["x"]		funcs: ["num_digits"]
	op: call 		dest: y_digits		type: int		args: ["y"]		funcs: ["num_digits"]
	op: eq   		dest: x_1_digit		type: bool		args: ["x_digits", "one"]	
	op: eq   		dest: y_1_digit		type: bool		args: ["y_digits", "one"]	
	op: or   		dest: cond		type: bool		args: ["x_1_digit", "y_1_digit"]	
	op: br   		args: ["cond"]		labels: ["then", "else"]
.then
	op: mul  		dest: result		type: int		args: ["x", "y"]	
	op: jmp  		labels: ["end"]
.else
	op: call 		dest: n 		type: int		args: ["x_digits", "y_digits"]		funcs: ["max"]
	op: div  		dest: n2		type: int		args: ["n", "two"]	
	op: call 		dest: pow_n2		type: int		args: ["base", "n2"]		funcs: ["pow"]
	op: div  		dest: a 		type: int		args: ["x", "pow_n2"]	
	op: call 		dest: b 		type: int		args: ["x", "pow_n2"]		funcs: ["mod"]
	op: div  		dest: c 		type: int		args: ["y", "pow_n2"]	
	op: call 		dest: d 		type: int		args: ["y", "pow_n2"]		funcs: ["mod"]
	op: call 		dest: z0		type: int		args: ["b", "d"]		funcs: ["karatsuba"]
	op: add  		dest: a_plus_b		type: int		args: ["a", "b"]	
	op: add  		dest: c_plus_d		type: int		args: ["c", "d"]	
	op: call 		dest: z1		type: int		args: ["a_plus_b", "c_plus_d"]		funcs: ["karatsuba"]
	op: call 		dest: z2		type: int		args: ["a", "c"]		funcs: ["karatsuba"]
	op: mul  		dest: hi_shift_factor		type: int		args: ["n2", "two"]	
	op: call 		dest: hi_shift		type: int		args: ["base", "hi_shift_factor"]		funcs: ["pow"]
	op: mul  		dest: hi_shifted		type: int		args: ["z2", "hi_shift"]	
	op: sub  		dest: mid		type: int		args: ["z1", "z2"]	
	op: sub  		dest: mid		type: int		args: ["mid", "z0"]	
	op: mul  		dest: mid_shifted		type: int		args: ["mid", "pow_n2"]	
	op: add  		dest: result		type: int		args: ["hi_shifted", "mid_shifted"]	
	op: add  		dest: result		type: int		args: ["result", "z0"]	
	op: jmp  		labels: ["end"]
.end
	op: ret  		args: ["result"]	
@num_digits
	op: const		dest: one		type: int		value: 1	
	op: const		dest: base		type: int		value: 10	
	op: id   		dest: digits		type: int		args: ["one"]	
	op: id   		dest: power		type: int		args: ["one"]	
.while.cond
	op: div  		dest: temp		type: int		args: ["n", "power"]	
	op: lt   		dest: is_term		type: bool		args: ["temp", "base"]	
	op: br   		args: ["is_term"]		labels: ["while.finish", "while.body"]
.while.body
	op: add  		dest: digits		type: int		args: ["digits", "one"]	
	op: mul  		dest: power		type: int		args: ["power", "base"]	
	op: jmp  		labels: ["while.cond"]
.while.finish
	op: ret  		args: ["digits"]	
@max
	op: gt   		dest: cond		type: bool		args: ["a", "b"]	
	op: br   		args: ["cond"]		labels: ["then", "else"]
.then
	op: ret  		args: ["a"]	
.else
	op: ret  		args: ["b"]	
@mod
	op: div  		dest: c 		type: int		args: ["a", "b"]	
	op: mul  		dest: d 		type: int		args: ["c", "b"]	
	op: sub  		dest: remainder		type: int		args: ["a", "d"]	
	op: ret  		args: ["remainder"]	
@pow
	op: id   		dest: v0		type: int		args: ["x"]	
	op: id   		dest: res		type: int		args: ["v0"]	
	op: const		dest: v2		type: int		value: 0	
	op: id   		dest: _i		type: int		args: ["v2"]	
.for.cond.1
	op: id   		dest: v3		type: int		args: ["_i"]	
	op: id   		dest: v4		type: int		args: ["n"]	
	op: const		dest: v5		type: int		value: 1	
	op: sub  		dest: v6		type: int		args: ["v4", "v5"]	
	op: lt   		dest: v7		type: bool		args: ["v3", "v6"]	
	op: br   		args: ["v7"]		labels: ["for.body.1", "for.end.1"]
.for.body.1
	op: id   		dest: v8		type: int		args: ["res"]	
	op: id   		dest: v9		type: int		args: ["x"]	
	op: mul  		dest: v10		type: int		args: ["v8", "v9"]	
	op: id   		dest: res		type: int		args: ["v10"]	
	op: id   		dest: v11		type: int		args: ["_i"]	
	op: const		dest: v12		type: int		value: 1	
	op: add  		dest: v13		type: int		args: ["v11", "v12"]	
	op: id   		dest: _i		type: int		args: ["v13"]	
	op: jmp  		labels: ["for.cond.1"]
.for.end.1
	op: id   		dest: v14		type: int		args: ["res"]	
	op: ret  		args: ["v14"]	
